veps/sig-network/ns-admin-nichotplug.md

```diff
@@ -0,0 +1,272 @@
+
+#### API Additions
+
+Currently, the CPU, memory and volume hotplug / hotunplug flows have a dedicated condition on the VMI object to mark the VMI
+for automatic migration:
+- HotVCPUChange
+- HotMemoryChange
+- VolumesChange
+
+Since we do not wish to introduce another resource-specific condition, a new generic allowed value will be added to `VirtualMachineInstanceConditionType`:
+
+```go
+VirtualMachineInstanceMigrationRequired VirtualMachineInstanceConditionType = "MigrationRequired"
+```
+
+For additional details, please see [Rational for Adding a Shared Condition](#appendix---rational-for-adding-a-shared-condition).
+
+#### Migration Based Hotplug / Hotunplug Flow
+
+1. The VM owner adds one or more new secondary vNICs and matching networks, or sets their state to `absent` to / on a running VM.
+2. The VM controller will:
+
+   - Not add the `RestartRequired` condition to the VM object.
+   - Patch the VMI object with the relevant vNICs and networks.
+3. The VMI controller will:
+   
+   - Patch the virt-launcher pod's Multus network attachment annotation (`k8s.v1.cni.cncf.io/networks`) - since the dynamic-networks-controller is not present, this will not have an affect.
+   - Update the VMI interface status.
+   - In case an SR-IOV NIC was hot-plugged - add the `MigrationRequired` condition to the VMI with its status set to `True` and a timestamp.
+   - Else - add the `MigrationRequired` condition to the VMI with its status set to `False` and a timestamp.
+4. On a future iteration of the VMI controller's reconciliation loop, the VMI controller will update the `MigrationRequired` condition status to `True` in case:
+
+   - The VMI has the `MigrationRequired` condition set with its status set to `False`.
+   - More than ten seconds had elapsed since the condition was set on the VMI (in order to allow the dynamic-networks-controller to operate).
+5. The Workload update controller will:
+   
+   - Identify the VMI as a candidate for automatic migration - since it has the `MigrationRequired` condition with its status set to `True`.
+   - Create a VMIM object for it.
+6. The Migration controller will remove the `MigrationRequired` condition once the migration is complete.
+7. The vNICs will be hot-plugged / hot-unplugged following the migration - same as today. 
+
+#### In-Place Hotplug / Hotunplug Flow
+
+1. The VM owner adds one or more new secondary vNICs with bridge binding and matching networks, or sets their state to `absent` to / on a running VM.
+2. The VM controller will:
+
+   - Not add the `RestartRequired` condition to the VM object.
+   - Patch the VMI object with the relevant vNICs and networks.
+3. The VMI controller will:
+
+    - Patch the virt-launcher pod's Multus network attachment annotation (`k8s.v1.cni.cncf.io/networks`).
+    - Update the VMI interface status.
+    - Add the `MigrationRequired` condition to the VMI with its status set to `False` and a timestamp.
```

Comment:
Would this be problematic in a future world where all of the hot plug implementations align to this single condition? Say you've also updated socket count at the same time which hot plug path wins setting `MigrationRequired` to their required value?

Reply:
I think there should be a precedence mechanism to determinate that.
NIC hotplug will still work if a migration will be performed while waiting for the dynamic-networks-controller.

Reply:
@lyarwood , I do not see any conflict, but if you detect one, could you share what it may be?

Any current or future hotplug implementation should reflect the request/need to trigger a migration through this condition. Therefore, if socket count joins the party, it’s logic should look similar to this (assuming no backoff required like with network):
- If no such condition type already in place, add it with status True.
- If condition type exists, set status to True (override anything that is there).

No one should override a status from True to False.

Reply:
So I wasn't aware that False to True wouldn't break the `dynamic-networks-controller` flow. Still True to False would be problematic. It's details like this that make be believe that a centralised VEP where this is all spelt out and documented for the different implementations is important.